New solver

B-head opened the pull request:
For now, release it with only minimal UI changes.

### Features:

- Support production loops.
- Support for bottom-up design.
- Support machine limit.
- Respects the list in order and sub-floor structure like the traditional solver.
- No need to specify free items.
- (Calculations are slow.)

### Not yet implemented:

- Production percentage.
  - Plan to implement instead an limits will be supported for items.
- Product prioritize.
  - Plan to implement instead an external product and external ingredient as virtual recipe.
  - It is also improvement of the free items feature.
- From the reference item, add a recipe that consumes it.
- Predicting the amount of machines needed when switching machines.

### Changes:

- The solver is now specified using the drop-down menu.
- The default solver for new subfactories is now the interior point solver.
  - Solvers for existing subfactories will not be changed.
- Add name and timescale to subfactory_data.
- Hide the percentage column when using matrix solver.
- Add and change locale strings.
- In recipe_line, the item button is displayed even if the amount is zero.

### Additional note:

- The solver has been modified to specify a reference instead of a production goal.
- Reference is the minimum amount of items to use in a sub-factory.
It is used to specify the items you want to produce or consume.

B-head committed:
Written a new LP solver.

B-head committed:
Using string.format().

B-head committed:
Include the subfactory name to subfactory_data.
Add dump of solver variables.

B-head committed:
Properly treat the same recipe.

B-head committed:
Add garbage.

B-head committed:
Reduce variables.

B-head committed:
Changed the initial value of vector s.
Set tolerance to MARGIN_OF_ERORR.

B-head committed:
Rename dot() to hadamard_product(). (I am an idiot.)
Expanded to allow calculation in matrix.

B-head committed:
Extremely small values were excluded from the step size calculation.

B-head committed:
Rewritten gaussian_elimination().

B-head committed:
Minor refactoring.

B-head committed:
Written SparseMatrix class.

B-head committed:
Eliminate variables in linear equations.

B-head committed:
Written cholesky_factorization().

B-head committed:
Support for selection of interior point solver.

B-head committed:
Adjust production_table for the new solver.

B-head committed:
Minor solver changes.

raiguard commented:
I will try this in my SE playthrough and let you know if I find anything.

raiguard commented:
![image](https://user-images.githubusercontent.com/3515394/135557784-5a195e82-b242-4b0d-8be3-c266dfd0489e.png)

Well it almost immediately crashes when I load it in my SE save. I tried removing it then re-enabling it but that doesn't help.

It must be crashing in `on_configuration_changed` somewhere, because it doesn't do this on a new game.

[The Final Frontier.zip](https://github.com/ClaudeMetz/FactoryPlanner/files/7263778/The.Final.Frontier.zip)

B-head committed:
Changed version to 1.1.26.

B-head commented:
Existing saves need to be migrated.
We have changed the version of json.info to 1.1.26, please try again.

B-head committed:
Fixed nil state trap.

B-head commented:
Fixed.
@raiguard Please try again.

B-head committed:
Fix: Crash when setting machine limit when reference item is terminated.

B-head committed:
Fix: Does not exit when the variable is empty.

B-head committed:
Force variables constraint.

B-head committed:
In recipe_line, the item button is displayed even if the amount is zero.

B-head committed:
Fix: Crash when ingredient_satisfaction is true.
Hide the fuel only ingredient if the amount is zero.

B-head committed:
Subfactories with loops to make it better default solution.

B-head committed:
Merge branch 'ClaudeMetz:master' into new_solver

B-head committed:
Rename migration file.

B-head committed:
Merge branch 'master' into new_solver

B-head committed:
Add LDoc to solver_util.lua

B-head committed:
Add a function to re-setup the meta table

B-head committed:
Modify orders

B-head committed:
Minor refactoring

B-head committed:
Add LDoc

B-head committed:
Refactoring name changes

B-head committed:
Fix LDoc

B-head committed:
Fix LDoc 2

B-head committed:
Improve performance
Use the previous solution as the initial value and calculate
from the point closest to the optimal solution.

B-head committed:
Fix LDoc 3

B-head committed:
Change to create and use an item ID

B-head committed:
Merge branch 'master' into new_solver

qdbp commented:
As far as I'm concerned, an LP solver is basically the raison d'etre of a factory planner for a pack like, e.g., Seablock. I would absolutely love to see this solver merged.

ClaudeMetz commented:
Yeah I agree it makes sense to have one, and I'm sorry I haven't really gotten to any of the proposals so far. I plan on working on the solver side of things at some point.
